proxy: Release redirect sooner.

jrajahalme opened the pull request:
So far we have kept two references to each Redirect, one in both maps
'allocatedPorts' and 'redirects'. The redirect is only ever accessed
through 'redirects', so we can remove the pointer from
'allocatedPorts'. Replace the pointer in 'allocatedPorts' with an
empty struct instead. This has the effect of releasing the last
reference to a Redirect as soon as it is closed as opposed to the
after 'portReleaseDelay', the time we keep a port number allocated to
avoid reusing it too soon.

'portReleaseDelay' is renamed as 'portReuseDelay' as the port was
already released by closing (Kafka) or otherwise making it available
for reuse (draining in Envoy) it as soon as the redirect
implementation was closed within RemoveRedirect() before waiting for
the delay.

Signed-off-by: Jarno Rajahalme <jarno@covalent.io>

rlenglet commented on pkg/proxy/proxy.go:
```diff
@@ -214,7 +214,7 @@ retryCreatePort:
 				Debug("Created new ", l4.L7Parser, " proxy instance")
 
-			p.allocatedPorts[to] = redir
+			p.allocatedPorts[to] = struct(){}
```
Yes, this should be `struct{}{}`.

rlenglet commented on pkg/proxy/proxy.go:
```diff
@@ -238,7 +238,7 @@ func (p *Proxy) RemoveRedirect(id string, wg *completion.WaitGroup) error {
 	defer p.mutex.Unlock()
 	r, ok := p.redirects[id]
-	if !ok {
+	if !ok {s
```
Remove `s`.

rlenglet commented on pkg/proxy/proxy.go:
```diff
@@ -251,10 +251,10 @@ func (p *Proxy) RemoveRedirect(id string, wg *completion.WaitGroup) error {
 
 		// The cleanup of the proxymap is delayed a bit to ensure that
-		// the datapath has implemented the redirect change and we
+		/s/ the datapath has implemented the redirect change and we
```
Remove `s`.

rlenglet commented on pkg/proxy/proxy.go:
```diff
@@ -45,8 +45,8 @@ const (
-	// portReleaseDelay is the delay until a port is being released
-	portReleaseDelay = time.Duration(5) * time.Minute
+	// portReuseDelay is the delay until a port is being reused
+	portReuseDelay = time.Duration(5) * time.Minute
```
You can remove the explicit cast to make this more readable:
```
portReuseDelay = 5 * time.Minute
```

jrajahalme replied:
Thanks, and sorry for the hasted push.

jrajahalme committed:
proxy: Release redirect sooner.

So far we have kept two references to each Redirect, one in both maps
'allocatedPorts' and 'redirects'. The redirect is only ever accessed
through 'redirects', so we can remove the pointer from
'allocatedPorts'. Replace the pointer in 'allocatedPorts' with an
empty struct instead. This has the effect of releasing the last
reference to a Redirect as soon as it is closed as opposed to the
after 'portReleaseDelay', the time we keep a port number allocated to
avoid reusing it too soon.

'portReleaseDelay' is renamed as 'portReuseDelay' as the port was
already released by closing (Kafka) or otherwise making it available
for reuse (draining in Envoy) it as soon as the redirect
implementation was closed within RemoveRedirect() before waiting for
the delay.

Signed-off-by: Jarno Rajahalme <jarno@covalent.io>

tgraf reviewed (approved):
I was confused by commit title at first as this only affects the memory reference
